src/functions/JpowerIntegrals.cpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * MRCPP, a numerical library based on multiresolution analysis and
+ * the multiwavelet basis which provide low-scaling algorithms as well as
+ * rigorous error control in numerical computations.
+ * Copyright (C) 2021 Stig Rune Jensen, Jonas Juselius, Luca Frediani and contributors.
+ *
+ * This file is part of MRCPP.
+ *
+ * MRCPP is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * MRCPP is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with MRCPP.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * For information on the complete list of contributors to MRCPP, see:
+ * <https://mrcpp.readthedocs.io/>
+ */
+
+#include "JpowerIntegrals.h"
+
+
+namespace mrcpp {
+
+
+JpowerIntegrals::JpowerIntegrals(double a, int scaling, int M, double treshold)
+{
+    this->scaling = scaling;
+    int N = 1 << scaling;
+    for(int l = 0; l < N; l++  )
+        integrals.push_back( calculate_J_power_inetgarls(l, a, M, treshold) );
+    for(int l = 1 - N; l < 0; l++  )
+        integrals.push_back( calculate_J_power_inetgarls(l, a, M, treshold) );
+}
+
+JpowerIntegrals::JpowerIntegrals(const JpowerIntegrals &other)
+{
+    scaling = other.scaling;
+
+    // Copy integrals (deep copy)
+    integrals.clear();
+    for (const auto& row : other.integrals)
+    {
+        std::vector<std::complex<double>> newRow(row);
+        integrals.push_back(newRow);
+    }    
+}
+
+JpowerIntegrals::~JpowerIntegrals()
+{
+    //for (auto& integral : integrals) 
+    //{
+    //    integral.clear();  
+    //}
+    // The loop is probably not needed for the complete memory deallocation
+    integrals.clear();
+}
+
+/// @brief in progress
+/// @param index - in progress
+/// @return in progress
+std::vector<std::complex<double>> & JpowerIntegrals::operator[](int index)
+{
+    if( index < 0 ) index += integrals.size();
+    return integrals[index];
+}
+
+std::vector<std::complex<double>> JpowerIntegrals::calculate_J_power_inetgarls(int l, double a, int M, double treshold)
+{
+    //const double pi = 3.1415926535897932384626433832795;
+    const std::complex<double> I(0.0, 1.0);  // Imaginary unit
+    const std::complex<double> J_0 = 0.25 * std::exp(-0.25 * I * M_PI) / std::sqrt(M_PI * a) * std::exp(0.25 * I * static_cast<double>(l * l) / a);
+    const std::complex<double> beta(0, 0.5 / a);
+    const std::complex<double> alpha = static_cast<double>(l) * beta;
```

Comment:
Few comments for these lines:
* These are local variables, so using `const` doesn't do much and, personally, I find it an unnecessary distraction.
* You can use `auto` as return type, if you want to save yourself some typing.
* You can use a literal to get the imaginary unit: https://en.cppreference.com/w/cpp/numeric/complex/operator%22%22i
```c++
using namespace std::complex_literals;

auto J_0 = 0.25 * std::exp(-0.25i * M_PI) / std::sqrt(M_PI * a) * std::exp(0.25i * static_cast<double>(l * l) / a);
```